File: src/C++/float/leading_minors.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-23 17:39:26
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-26 09:05:49
@FilePath: \asm_matrix_benchmark\src\C++\float\leading_minors.cpp
@Description: leading principal minors c++ code
*************************************************************/

#include "matrix_float_cpp.h"
#include <stdexcept>

/***********************************************************
@description: Computes all leading principal minors of the matrix.
@return {*}
*************************************************************/
std::vector<MatrixFloat> MatrixFloat::leading_minors() const
{
    if (rows != cols)
    {
        throw std::invalid_argument("leading_minors: square matrix required");
    }

    std::vector<MatrixFloat> result;
    result.reserve(rows);

    // Extract leading principal submatrices of order 1..rows
    // Each submatrix is the top-left corner of the original matrix
    for (size_t order = 1; order <= rows; order++)
    {
        MatrixFloat sub(order, order);
        for (size_t i = 0; i < order; i++)
        {
            for (size_t j = 0; j < order; j++)
            {
                sub(i, j) = data[i * cols + j];
            }
        }
        result.push_back(std::move(sub));
    }
    return result;
}

File: src/JavaScript/leading_minors.js
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 20:10:32
@LastEditors: GitHubCopilot
@LastEditTime: 2026-06-29 20:10:32
@FilePath: \asm_matrix_benchmark\src\JavaScript\leading_minors.js
@Description: leading minors JavaScript code
*************************************************************/

"use strict";

const { checkMatrix, buildTypedMatrix } = require("./common");

/**
 * Build the top-left k x k principal submatrix.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {number} k
 * @param {"int"|"float"|"double"} type
 * @returns {{rows:number, cols:number, data:number[]}|null}
 */
function topLeftMinor(m, k, type) {
  const data = new Array(k * k);
  for (let i = 0; i < k; i++) {
    for (let j = 0; j < k; j++) {
      data[i * k + j] = m.data[i * m.cols + j];
    }
  }
  return buildTypedMatrix(k, k, data, type);
}

/**
 * Return all leading principal minors as matrix list:
 * [1x1, 2x2, ..., nxn].
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {"int"|"float"|"double"} type
 * @returns {Array<{rows:number, cols:number, data:number[]}>|null}
 */
function getLeadingMinors(m, type) {
  if (!checkMatrix(m, "m")) {
    return null;
  }
  if (m.rows !== m.cols) {
    console.error("Matrix must be square!");
    return null;
  }

  const list = [];
  for (let k = 1; k <= m.rows; k++) {
    const minor = topLeftMinor(m, k, type);
    if (minor === null) {
      return null;
    }
    list.push(minor);
  }
  return list;
}

function get_leading_minors_int(m) {
  return getLeadingMinors(m, "int");
}

function get_leading_minors_float(m) {
  return getLeadingMinors(m, "float");
}

function get_leading_minors_double(m) {
  return getLeadingMinors(m, "double");
}

module.exports = {
  get_leading_minors_int,
  get_leading_minors_float,
  get_leading_minors_double,
};
